keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1023,26 +1024,125 @@ this enhancement.
 
 #### Enforcement Actions
 
-For phase 1, all violations implicitly result in a `deny` enforcement action.
+For parity with admission webhooks, a validating policy may also emit audit
+annotations and warnings:
 
-For phase 2, we intend to support multiple enforcement actions.
+- [Audit
+  annotations](https://github.com/kubernetes/kubernetes/blob/97bbf07d3f3f20332912ee411fdf75ce84425e28/staging/src/k8s.io/api/admission/v1/types.go#L142)
+  are key/value pairs included in the audit event for an admission request. The
+  audit annotation key supplied by the policy definition author will be prefixed
+  with the name of the `ValidatingAdmissionPolicy` and policy binding, e.g.:
+  `mypolicy.mygroup.example.com/mybinding.mygroup.example.com/<annotation-key>`.
+- [Warnings](https://kubernetes.io/blog/2020/09/03/warnings/#admission-webhooks)
+  are string messages that are returned to API clients. Warning are returned for
+  both requests that are accepted and requests that are rejected.
 
-Use cases:
+`ValidatingAdmissionPolicy` may declare audit annotations in the policy
+definition. E.g.:
 
-- Cluster admin would like to rollout a policies, sometimes in bulk, without
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  ...
+  validations:
+    - expression: <expression>
+  auditAnnotations:
+    - includeWhen: <expression> # optional field
+      key: "my-audit-key"
+      valueExpression: <expression that evaluates to a string>
```

Comment:
are we at all concerned about the cost of 1-2 additional expressions per audit annotation?

Reply:
The `valueExpression` seems reasonable.  `includeWhen` might not be justified.  We could simplify this by omitting the annotation if the `valueExpressions` returns a specific value. Maybe we allow the expression to evaluate to either a string or null, and if it evaluates to null, we don't write emit audit annotation.